src/main/java/org/spongepowered/api/resource/Resource.java

```diff
@@ -0,0 +1,245 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.resource;
+
+import com.google.common.io.ByteStreams;
+import com.google.common.io.CharStreams;
+import com.google.common.io.MoreFiles;
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.data.DataView;
+import org.spongepowered.api.data.persistence.DataFormat;
+import org.spongepowered.api.data.persistence.InvalidDataFormatException;
+import org.spongepowered.api.util.ResettableBuilder;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.Reader;
+import java.nio.charset.Charset;
+import java.nio.file.OpenOption;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.WillNotClose;
+
+/**
+ * A resource can represent any kind of loaded data. It can be a file on the
+ * filesystem, a network location, or even generated at runtime. Use
+ * {@link #openStream()} to load the data held by a resource.
+ */
+public interface Resource extends InputStreamSupplier {
+
+    /**
+     * Creates a new builder for a {@link Resource}.
+     *
+     * @return A new builder
+     */
+    static Builder builder() {
+        return Sponge.getRegistry().createBuilder(Builder.class);
+    }
+
+    /**
+     * Creates a new resource for the given {@link ResourcePath} and
+     * {@link InputStream} supplier. The supplier should return a new
+     * input stream each time it is called.
+     *
+     * @param path The path of the resource
+     * @param stream Supplier for the input stream
+     * @return
+     */
+    static Resource of(ResourcePath path, InputStreamSupplier stream) {
+        return builder().path(path).stream(stream).build();
+    }
+
+    /**
+     * Gets the path of this resource.
+     *
+     * @return The path
+     */
+    ResourcePath getResourcePath();
+
+    /**
+     * Gets the {@link Pack} which owns this resource. The pack is set
+     * automatically when the resource is added to a pack.
+     *
+     * @return The parent pack.
+     */
+    Pack getPack();
```

Comment:
Do Resources always have Packs?

Reply:
I think there may be a default pack for minecraft

Reply:
If used properly, there are no orphans in the resource manager. The only way for this to return null is for the developer to use `Resource.builder()` and not add it to a pack.

I'm actually working on something that makes this a bit more foolproof. Standby.